### PR TITLE
[filter/snpe] Support "user supplied buffer" for snpe

### PR DESCRIPTION
**Pull request by anyj0527**
- Support "userbuffer" for snpe filter
- With this mode, the memcpy from/into model's input/output does not occur
- Note that it only supports float type
- Add two simple cases to test user buffer execution

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 1]
```diff
@@ -574,8 +698,8 @@ snpe_subplugin::init_filter_snpe (void)
       "Layer names for the output, separated by ';'. E.g., 'layer0;layer1;layer2'",
       "InputType",
       "Set the data type of the input {'float32 (default)', 'uint8'}",
-      "OutputType",
-      "Set the data type of the output {'float32 (default)', 'uint8'}",
+      "UserBuffer",
+      "Use user supplied buffers for input/output tensors {'false (default)', 'true'}",
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:702]: Output Type is removed??

Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc:702]: Thanks, it was my mistake.

**Review by again4you — COMMENTED** [review]
First of all, your work will improve the inference performance because of removing the memory copy. After reviewing your code, I added some questions and my opinions. :)

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 2]
```diff
@@ -390,40 +406,111 @@ snpe_subplugin::configure_instance (const GstTensorFilterProperties *prop)
...
-  snpe->execute (input_tensor_map, output_tensor_map);
+      if (user_input_buffers.back () == nullptr) {
+        std::cerr << "Error while creating user buffer." << std::endl;
+        return;
+      }
 
-  setTensorProp (inputInfo, input_tensor_map, input_data_type);
-  setTensorProp (outputInfo, output_tensor_map, output_data_type);
+      input_buffer_map.add (name, user_input_buffers.back ().get ());
+    }
```
[the remark this answers is not in the document]
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: Cool. I've refactored the code block into a member function and now it looks better. Thanks!

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 3]
```diff
@@ -441,48 +528,60 @@ snpe_subplugin::invoke (const GstTensorMemory *input, GstTensorMemory *output)
...
+          break;
+        }
+        case _NNS_UINT8: {
+          uint8_t *inbuf = (uint8_t *) input[i].data;
+          std::copy (inbuf, inbuf + fsize, input_tensors[i].get ()->begin ());
+          break;
+        }
+        default:
+          throw std::runtime_error ("Got invalid input data type");
+          break;
```
[the remark this answers is not in the document]
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: Right. I removed the needless line.

File: ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc [hunk 4]
```diff
@@ -390,40 +406,111 @@ snpe_subplugin::configure_instance (const GstTensorFilterProperties *prop)
...
+        bufSize *= bufferShape[i];
+      }
+
+      std::unique_ptr<zdl::DlSystem::UserBufferEncoding> userBufferEncoding;
+      userBufferEncoding = std::unique_ptr<zdl::DlSystem::UserBufferEncodingFloat>(new zdl::DlSystem::UserBufferEncodingFloat ());
+      user_output_buffers.push_back (ubFactory.createUserBuffer (NULL, bufSize, strides, userBufferEncoding.get ()));
+
+      if (user_output_buffers.back () == nullptr) {
+        std::cerr << "Error while creating user buffer." << std::endl;
+        return;
```
[the remark this answers is not in the document]
Comment (anyj0527) [reply ext/nnstreamer/tensor_filter/tensor_filter_snpe.cc]: I've updated "the error log and simple return" with throw exceptions. Thanks for pointing. PTAL

**Review by myungjoo — APPROVED** [review]
LGTM

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**